src/transformStateWithClones.js

```diff
@@ -7,7 +7,40 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let stateCopy = { ...state };
+  const stateHistory = [];
+
+  for (let index = 0; index < actions.length; index++) {
+    const action = actions[index];
+    const type = action.type;
+
+    switch (type) {
+      case 'addProperties':
+        stateCopy = { ...stateCopy, ...action.extraData };
+        stateHistory.push({ ...stateCopy });
+        break;
+
+      case 'removeProperties':
+        for (let i = 0; i < action.keysToRemove.length; i++) {
+          delete stateCopy[action.keysToRemove[i]];
+        }
+        stateHistory.push({ ...stateCopy });
+        break;
+
+      case 'clear':
+        const keys = Object.keys(stateCopy);
+
+        for (let i = 0; i < keys.length; i++) {
+          delete stateCopy[keys[i]];
+        }
+        stateHistory.push({});
+        break;
+      default:
+        stateHistory.push({ error: 'type-error' });
```

Comment:
The `default` case in a `switch` statement should ideally handle unexpected cases. Pushing an error directly to `stateHistory` may not be the best approach as it alters the expected structure of the history array. It would be better to throw an error or handle the unexpected `type` in a way that does not interfere with the state history structure.